src/ReservoirPair.sol

```diff
@@ -148,15 +148,45 @@ abstract contract ReservoirPair is IAssetManagedPair, ReservoirERC20 {
         require(aBalance0 <= type(uint104).max && aBalance1 <= type(uint104).max, "RP: OVERFLOW");
         require(aReserve0 <= type(uint104).max && aReserve1 <= type(uint104).max, "RP: OVERFLOW");
 
-        uint32 lBlockTimestamp = uint32(_currentTime());
+        uint32 lBlockTimestamp = _currentTime();
         uint32 lTimeElapsed;
         unchecked {
             // underflow is desired
             // however in the case where no swaps happen in ~68 years (2 ** 31 seconds) the timeElapsed would underflow twice
             lTimeElapsed = lBlockTimestamp - aBlockTimestampLast;
         }
-        if (lTimeElapsed > 0 && aReserve0 != 0 && aReserve1 != 0) {
-            _updateOracle(aReserve0, aReserve1, lTimeElapsed, lBlockTimestamp);
+
+        // both balance should never be zero
+        // but necessary to check so we don't pass 0 values into arithmetic operations
+        if (aBalance0 != 0 && aBalance1 != 0) {
+            Observation storage lPrevious = _observations[_slot0.index];
+            (uint256 lInstantRawPrice, int256 lLogInstantRawPrice) = _calcSpotAndLogPrice(aBalance0, aBalance1);
+            // checks to make sure we don't create a new oracle sample at the block of creation
+            if (lTimeElapsed > 0 && aReserve0 != 0 && aReserve1 != 0) {
+                (, int256 lLogInstantClampedPrice) = _calcClampedPrice(
+                    lInstantRawPrice,
+                    lLogInstantRawPrice,
+                    LogCompression.fromLowResLog(lPrevious.logInstantClampedPrice),
+                    lTimeElapsed,
+                    aBlockTimestampLast
+                );
+                _updateOracleNewSample(
+                    lPrevious, lLogInstantRawPrice, lLogInstantClampedPrice, lTimeElapsed, lBlockTimestamp
+                );
+            } else {
+                // for instant price updates in the same timestamp, we use the time difference from the previous oracle observation as the time elapsed
+                lTimeElapsed = lBlockTimestamp - lPrevious.timestamp;
+
+                (, int256 lLogInstantClampedPrice) = _calcClampedPrice(
+                    lInstantRawPrice,
+                    lLogInstantRawPrice,
+                    LogCompression.fromLowResLog(lPrevious.logInstantClampedPrice),
+                    lTimeElapsed,
+                    lPrevious.timestamp
+                );
+
+                _updateOracleInstantPrices(lPrevious, lLogInstantRawPrice, lLogInstantClampedPrice);
+            }
```

Comment:
We have to run this every time because we want to store the latest instant price in the oracle slot right? Is it not sufficient to only update instant price with rest of oracle inputs?